tracing-subscriber/src/registry/fmt.rs

```diff
@@ -0,0 +1,315 @@
+                    &mut b
+                }
+            };
+
+            if (self.is_interested)(event) {
+                let ctx = self.make_ctx(ctx);
+                if self.fmt_event.format_event(&ctx, &mut buf, event).is_ok() {
+                    let mut writer = self.make_writer.make_writer();
+                    let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+                }
+            }
+
+            buf.clear();
+        });
+    }
+}
+
+/// `FmtContext` is used to propogate subscriber context to tracing_subscriber::fmt.
+pub struct FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    pub(crate) ctx: Context<'a, S>,
+    pub(crate) fmt_fields: &'a N,
+}
+
+impl<'a, S, N> fmt::Debug for FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FmtContext").finish()
+    }
+}
+
+impl<'a, S, N> FormatFields<'a> for FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    fn format_fields<R: RecordFields>(
+        &self,
+        writer: &'a mut dyn fmt::Write,
+        fields: R,
+    ) -> fmt::Result {
+        self.fmt_fields.format_fields(writer, fields)
+    }
+}
+
+impl<'a, S, N> FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    // TODO(david): consider an alternative location for this.
+    /// Visits parent spans. Used to visit parent spans when formatting spans
+    /// and events
+    pub fn visit_spans<E, F>(&self, f: F) -> Result<(), E>
+    where
+        F: FnMut(&Id) -> Result<(), E>,
+    {
+        let current_span = self.ctx.current_span();
+        let id = match current_span.id() {
+            Some(id) => id,
+            None => return Ok(()),
+        };
+        let span = match self.ctx.span(id) {
+            Some(span) => span,
+            None => return Ok(()),
+        };
+
+        // this isn't _great_, because we allocate for each iteration. a
+        // better way to handle this would be to the recursive approach that
+        // `fmt` pursues that _does not_ entail any allocation in this fmt'ing
+        // spans path. however, that requires passing the store to `visit_spans`
+        // with a different lifetime, and i'm too lazy to sort that out now. this
+        // workaround shouldn't remaining in the final shipping version _unless_
+        // benchmarks show that allocating two vecs is preferable to not-very-deep
+        // recursion. I'd be suprised if that's the case.
+        let mut current: SmallVec<[Id; 16]> = smallvec![id.clone()];
```

Comment:
I think using a `SmallVec` here is a great idea, in most cases, we still won't allocate, and if we do, it would be in a case where the previous impl would do a _lot_ of recursion. So I think this is a good call.

However, we have a crate feature flag for the use of smallvec in filter, to allow disabling the dependency. I think this needs to honor that feature flag. I'd consider adding a type alias like

```rust
#[cfg(feature = "smallvec")]
type IdVec = smallvec::SmallVec<[Id; 16]>;
#[cfg(not(feature = "smallvec"))]
type IdVec = Vec<Id>;

Reply:
That's a good point; added.